File: Data Structure/String/Medium/LET_no1249_Minimum Remove to Make Valid Parentheses.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        stack = []
        removes = set()
        for i, c in enumerate(s):
            if c == '(':
                stack.append(i)
            elif c == ')':
                if stack:
                    stack.pop()
                else:
                    removes.add(i)
        
        for i in stack:
            removes.add(i)
            
        return ''.join(c for i, c in enumerate(s) if i not in removes)

File: Data Structure/String/Medium/LET_no1249_Minimum Remove to Make Valid Parentheses.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        vector<int> memo;
        unordered_set<int> removes;
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(')
                memo.push_back(i);
            else if (s[i] == ')') {
                if (!memo.empty())
                    memo.pop_back();
                else
                    removes.insert(i);
            }
        }

        for (int i : memo)
            removes.insert(i);

        string res;
        for (int i = 0; i < s.size(); i++) {
            if (!removes.count(i))
                res.push_back(s[i]);
        }

        return res;
    }
};